src/jit/hwintrinsiccodegenxarch.cpp

```diff
@@ -2008,6 +2011,115 @@ void CodeGen::genAvxOrAvx2Intrinsic(GenTreeHWIntrinsic* node)
             break;
         }
 
+        case NI_AVX2_GatherVector128:
+        case NI_AVX2_GatherVector256:
+        case NI_AVX2_GatherMaskVector128:
+        case NI_AVX2_GatherMaskVector256:
+        {
+            GenTreeArgList* list = op1->AsArgList();
+            op1                  = list->Current();
+            list                 = list->Rest();
+            op2                  = list->Current();
+            list                 = list->Rest();
+            GenTree* op3         = list->Current();
+            list                 = list->Rest();
+            GenTree* op4         = nullptr;
+            GenTree* lastOp      = nullptr;
+            GenTree* indexOp     = nullptr;
+
+            regNumber op3Reg       = REG_NA;
+            regNumber op4Reg       = REG_NA;
+            regNumber addrBaseReg  = REG_NA;
+            regNumber addrIndexReg = REG_NA;
+            regNumber maskReg      = node->ExtractTempReg(RBM_ALLFLOAT);
+
+            op1Reg = op1->gtRegNum;
+            genConsumeRegs(op1);
+            op2Reg = op2->gtRegNum;
+            genConsumeRegs(op2);
+
+            if (intrinsicId == NI_AVX2_GatherMaskVector128 || intrinsicId == NI_AVX2_GatherMaskVector256)
+            {
+                op4    = list->Current();
+                list   = list->Rest();
+                lastOp = list->Current();
+                op3Reg = op3->gtRegNum;
+                genConsumeRegs(op3);
+                op4Reg = op4->gtRegNum;
+                genConsumeRegs(op4);
+                addrBaseReg  = op2Reg;
+                addrIndexReg = op3Reg;
+                indexOp      = op3;
+
+                // the mask register will be cleared by gather instructions
+                emit->emitIns_R_R(INS_movaps, attr, maskReg, op4Reg);
+
+                if (targetReg != op1Reg)
+                {
+                    // copy source vector to the target register for masking merge
+                    emit->emitIns_R_R(INS_movaps, attr, targetReg, op1Reg);
+                }
+            }
+            else
+            {
+                addrBaseReg  = op1Reg;
+                addrIndexReg = op2Reg;
+                indexOp      = op2;
+                lastOp       = op3;
+
+                // generate all-one mask vector
+                emit->emitIns_SIMD_R_R_R(INS_pcmpeqd, attr, maskReg, maskReg, maskReg);
+            }
+
+            regNumber lastOpReg = lastOp->gtRegNum;
+            genConsumeRegs(lastOp);
```

Comment:
What is the purpose of having this here, rather than always consuming op3 before the if-stmt above, and then taking care of the other sources in that if stmt?